src/NonfungiblePositionManager.sol

```diff
@@ -0,0 +1,226 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.24;
+
+import {IPoolManager} from "@uniswap/v4-core/src/interfaces/IPoolManager.sol";
+import {PoolKey} from "@uniswap/v4-core/src/types/PoolKey.sol";
+import {PoolIdLibrary} from "@uniswap/v4-core/src/types/PoolId.sol";
+import {Currency, CurrencyLibrary} from "@uniswap/v4-core/src/types/Currency.sol";
+import {BalanceDelta} from "@uniswap/v4-core/src/types/BalanceDelta.sol";
+import {SafeCast} from "@uniswap/v4-core/src/libraries/SafeCast.sol";
+import {StateLibrary} from "@uniswap/v4-core/src/libraries/StateLibrary.sol";
+import {TransientStateLibrary} from "@uniswap/v4-core/src/libraries/TransientStateLibrary.sol";
+
+import {ERC721Permit} from "./base/ERC721Permit.sol";
+import {INonfungiblePositionManager, Actions} from "./interfaces/INonfungiblePositionManager.sol";
+import {SafeCallback} from "./base/SafeCallback.sol";
+import {ImmutableState} from "./base/ImmutableState.sol";
+import {Multicall} from "./base/Multicall.sol";
+import {PoolInitializer} from "./base/PoolInitializer.sol";
+import {CurrencySettleTake} from "./libraries/CurrencySettleTake.sol";
+import {LiquidityRange, LiquidityRangeId, LiquidityRangeIdLibrary} from "./types/LiquidityRange.sol";
+
+contract NonfungiblePositionManager is
+    INonfungiblePositionManager,
+    ERC721Permit,
+    PoolInitializer,
+    Multicall,
+    SafeCallback
+{
+    using CurrencyLibrary for Currency;
+    using CurrencySettleTake for Currency;
+    using PoolIdLibrary for PoolKey;
+    using LiquidityRangeIdLibrary for LiquidityRange;
+    using StateLibrary for IPoolManager;
+    using TransientStateLibrary for IPoolManager;
+    using SafeCast for uint256;
+
+    /// @dev The ID of the next token that will be minted. Skips 0
+    uint256 public nextTokenId = 1;
+
+    // maps the ERC721 tokenId to its Range (poolKey, tick range)
+    mapping(uint256 tokenId => LiquidityRange range) public tokenRange;
+
+    constructor(IPoolManager _manager)
+        ImmutableState(_manager)
+        ERC721Permit("Uniswap V4 Positions NFT-V1", "UNI-V4-POS", "1")
+    {}
+
+    /// @param unlockData is an encoding of actions, params, and currencies
+    /// @return returnData is the endocing of each actions return information
+    function modifyLiquidities(bytes calldata unlockData, uint256 deadline)
```

Comment:
reminder of TODO to add a reentrancy lock

Reply:
https://github.com/Uniswap/v4-periphery/issues/170